state/protocol/badger/mutator.go

```diff
@@ -119,8 +138,8 @@ func (m *FollowerState) Extend(ctx context.Context, candidate *flow.Block) error
 		return fmt.Errorf("payload seal(s) not compliant with chain state: %w", err)
 	}
 
-	// insert the block and index the last seal for the block
-	err = m.insert(ctx, candidate, last)
+	// insert the block, certifying QC and index the last seal for the block
+	err = m.insert(ctx, candidate, certifyingQC, last)
 	if err != nil {
```

Comment:
```suggestion
	if err != nil {
		if state.IsInvalidExtensionError(err) {
			// since we have a QC for this block, it cannot be an invalid extension
			return fmt.Errorf("unexpected invalid block (id=%x) with certifying qc (id=%x): %v", candidate.ID(), certifyingQC.ID(), err)
```
Suggesting to explicitly convert this to an exception.

Reply:
this wouldn't work since we just wrap the error, but I agree, invalid extension should be threaten as explicit exception

Reply:
https://github.com/onflow/flow-go/pull/3949/commits/a42afb0d3ad5d09d473b81730b0c13e41fd310b8

Reply:
Thank you 🙇. `%s` works just as well, but for reference, `%v` does not wrap the error like `%w`. `...%v", err` and `...%s", err.Error()` behave the same way: https://go.dev/play/p/bQEVCHgHOY5

Reply:
I think using `...%v", err` is quite a risky thing. I wouldn't be surprised if people misstook that as a typo. Would prefer the explicit `Exception` type. Makes it very specific in the code reducing the need for detailed documentation.

Reply:
I am a bit conflicted here. This is because we don't document `InvalidExtensionError` as an expected error (because it isn't), so I feel this would fall under the category "anything other than the explicitly documented sentinels are exceptions".

Reply:
> I am a bit conflicted here. This is because we don't document `InvalidExtensionError` as an expected error (because it isn't), so I feel this would fall under the category "anything other than the explicitly documented sentinels are exceptions".

Right, but I still think that it's worth explicitly returning an untyped exception instead of using a sentinel error. I think that it makes sense to return sentinels only if it's an expected return value. We can continue this discussion, for know no harm in doing it this way, we can always change it back.